lib/evm/instruction/impl.ex

```diff
@@ -1642,19 +1622,35 @@ defmodule EVM.Instruction.Impl do
   # Helper function to push to the stack within machine_state.
   @spec push(MachineState.t | Stack.t, EVM.val) :: op_result
   defp push(machine_state=%MachineState{}, val) do
-    %{machine_state | stack: machine_state.stack |> Stack.push(val|>mod)}
+    %{
+      machine_state |
+      stack: machine_state.stack
+        |> Stack.push(val |> encode_signed)
+    }
   end
 
   # Helper function to just return an updated stack
   defp push(stack, val) when is_list(stack) do
-    %{stack: Stack.push(stack, val|>mod)}
+    %{stack: Stack.push(stack, val |> encode_signed)}
   end
 
-  @spec mod(integer()) :: EVM.val
-  defp mod(n), do: rem(n, EVM.max_int())
+  @spec wrap_int(integer()) :: EVM.val
+  defp wrap_int(n) when n > 0, do: band(n, EVM.max_int() - 1)
+  defp wrap_int(n), do: n
+
 
   # TODO: signed?
   @spec decode(binary()) :: EVM.val
-  defp decode(bin), do: :binary.decode_unsigned(bin) |> mod
+  defp decode(bin), do: :binary.decode_unsigned(bin) |> wrap_int
+
+  def decode_signed(n) do
+    <<sign :: size(1), _ :: bitstring>> = :binary.encode_unsigned(n)
```

Comment:
This is two's complement? can't we just use `:binary.decode_signed(n)`?

Reply:
I was hoping erlang would have had that but I didn't see it 🤔

    iex(2)> :binary.decode_signed(<<1>>)
    ** (UndefinedFunctionError) function :binary.decode_signed/1 is undefined